encoding/graph/common/schema_id_allocator.rs

```diff
@@ -0,0 +1,139 @@
+    graph::{
+        definition::definition_key::{DefinitionID, DefinitionKey},
+        type_::{
+            vertex::{TypeID, TypeIDUInt, TypeVertex},
+            Kind,
+        },
+        Typed,
+    },
+    layout::prefix::Prefix,
+    Keyable,
+};
+
+pub type TypeVertexAllocator = SchemaIDAllocator<TypeVertex<'static>>;
+pub type DefinitionKeyAllocator = SchemaIDAllocator<DefinitionKey<'static>>;
+
+pub trait AllocationHelper: Sized {
+    const MIN_ID: u64;
+    const MAX_ID: u64;
+
+    fn object_from_id(prefix: Prefix, id: u64) -> Self;
+    fn id_from_key<'b>(key: StorageKey<'b, BUFFER_KEY_INLINE>) -> u64;
+    fn ids_exhausted_error(prefix: Prefix) -> EncodingError;
+}
+
+pub struct SchemaIDAllocator<T: AllocationHelper> {
+    last_allocated_type_id: AtomicU64,
+    prefix: Prefix,
+    phantom: PhantomData<T>,
+}
+
+impl<'a, T: AllocationHelper + Keyable<'a, BUFFER_KEY_INLINE>> SchemaIDAllocator<T> {
+    pub fn new(prefix: Prefix) -> Self {
+        Self { last_allocated_type_id: AtomicU64::new(0), prefix, phantom: PhantomData }
+    }
+
+    fn find_unallocated_id<Snapshot: WritableSnapshot>(
+        &self,
+        snapshot: &mut Snapshot,
+        start: u64,
+    ) -> Result<Option<u64>, EncodingError> {
+        let mut schema_object_iter = snapshot.iterate_range(KeyRange::new_inclusive(
+            T::object_from_id(self.prefix, start).into_storage_key(),
+            T::object_from_id(self.prefix, T::MAX_ID).into_storage_key(),
+        ));
+        for expected_next_id in start..=T::MAX_ID {
+            match schema_object_iter.next() {
+                None => return Ok(Some(expected_next_id)),
+                Some(Err(err)) => {
+                    return Err(EncodingError::SchemaIDAllocate { source: err.clone() });
+                }
+                Some(Ok((actual_next_key, _))) => {
+                    let actual_id = T::id_from_key(actual_next_key);
+                    if actual_id != expected_next_id {
+                        return Ok(Some(expected_next_id));
+                    }
+                }
+            }
+        }
+        Ok(None)
+    }
+
+    pub fn allocate<Snapshot: WritableSnapshot>(&self, snapshot: &mut Snapshot) -> Result<T, EncodingError> {
+        let found = self.find_unallocated_id(snapshot, self.last_allocated_type_id.load(Relaxed))?;
+        if let Some(allocated_id) = found {
+            self.last_allocated_type_id.store(allocated_id.clone(), Relaxed);
+            Ok(T::object_from_id(self.prefix, allocated_id))
+        } else {
+            match self.find_unallocated_id(snapshot, 0)? {
+                None => Err(T::ids_exhausted_error(self.prefix)),
+                Some(allocated_id) => {
+                    self.last_allocated_type_id.store(allocated_id.clone(), Relaxed);
+                    Ok(T::object_from_id(self.prefix, allocated_id))
+                }
+            }
+        }
+    }
+}
+
+impl AllocationHelper for TypeVertex<'static> {
```

Comment:
I think this is approaching correct, but I'm not sure it's there yet.
My first instinct is to call this `impl SchemaID for TypeVertex`. 
However, now it makes me think of the overlap with `TypeID`... is `TypeID` redundant? Somehow these are now very similar ideas.
And secondly, calling it `SchemaID` lets us move the implementation into `TypeVertex`, which is I think where this belongs (and below in DefinitionKey).